lib/libzpool/uzfs_io.c

```diff
@@ -402,3 +424,35 @@ uzfs_read_metadata(zvol_state_t *zv, char *buf, uint64_t offset, uint64_t len,
 
 	return (ret);
 }
+
+/*
+ * Update metadata granularity. This is done when zvol is opened the first time
+ * (based on value sent from iSCSI target) and afterwards the granularity must
+ * not change.
+ */
+int
+uzfs_update_metadata_granularity(zvol_state_t *zv, uint64_t block_size)
+{
+	int error;
+	dmu_tx_t *tx;
+
+	if (block_size == zv->zv_metavolblocksize)
+		return (0);	/* nothing to update */
+	if (block_size > zv->zv_volblocksize)
```

Comment:
below two constraints are not needed, if I'm not wrong.

Reply:
I did it wrong I think :) I think it should be
```
if (tgt_block_size < zv->zv_volblocksize)
...
if (tgt_block_size % zv->zv_volblocksize)
...
```
The reasoning is that if target uses smaller block size at zvol level it will lead to inefficient write/read as we always have to read/write more data than the target requested. On the other hand it does not seem to cause any problem - it only affects performance. So should I remove those checks to simplify the code?

Reply:
I think - lets not add constraints in data plane. Probably, this can be something for control plane to give warning or so.